dhcp/dhcp.go

```diff
@@ -0,0 +1,365 @@
+func (c *DHCP) sendDHCPDiscover(fd int, packet []byte) error {
+	raddr := &net.UDPAddr{IP: net.IPv4bcast, Port: dhcpServerPort}
+	laddr := &net.UDPAddr{IP: net.IPv4zero, Port: dhcpClientPort}
+	var destination [net.IPv4len]byte
+	copy(destination[:], raddr.IP.To4())
+
+	packetBytes, err := MakeRawUDPPacket(packet, *raddr, *laddr)
+	if err != nil {
+		return errors.Wrap(err, "error making raw udp packet")
+	}
+
+	// Create sockaddr_ll structure for sending the packet
+	remoteAddr := unix.SockaddrInet4{Port: laddr.Port, Addr: destination}
+
+	// Send the packet using the raw socket
+	return errors.Wrap(unix.Sendto(fd, packetBytes, 0, &remoteAddr), "failed unix send to")
+}
+
+// Receive DHCP response packet using unix.Recvfrom
+// returns nil if a reply is received with the xid, otherwise an error is returned
+func (c *DHCP) receiveDHCPResponse(fd int, xid TransactionID) error {
+	recvErrors := make(chan error, 1)
+	go func(errs chan<- error) {
+		// set read timeout
+		timeout := unix.NsecToTimeval(DefaultReadTimeout.Nanoseconds())
+		if innerErr := unix.SetsockoptTimeval(fd, unix.SOL_SOCKET, unix.SO_RCVTIMEO, &timeout); innerErr != nil {
+			errs <- innerErr
+			return
+		}
+		// loop will only exit if there is an error, or we find our reply packet
+		for {
+			buf := make([]byte, MaxUDPReceivedPacketSize)
+			n, _, innerErr := unix.Recvfrom(fd, buf, 0)
+			if innerErr != nil {
+				errs <- innerErr
+				return
+			}
+			// check header
+			var iph ipv4.Header
+			if err := iph.Parse(buf[:n]); err != nil {
+				// skip non-IP data
+				continue
+			}
+			if iph.Protocol != udpProtocol {
+				// skip non-UDP packets
+				continue
+			}
+			udph := buf[iph.Len:n]
+			// source is from dhcp server if receiving
+			srcPort := int(binary.BigEndian.Uint16(udph[0:2]))
+			if srcPort != dhcpServerPort {
+				continue
+			}
+			// client is to dhcp client if receiving
+			dstPort := int(binary.BigEndian.Uint16(udph[2:4]))
+			if dstPort != dhcpClientPort {
+				continue
+			}
+			// check payload
+			pLen := int(binary.BigEndian.Uint16(udph[4:6]))
+			payload := buf[iph.Len+8 : iph.Len+pLen]
+
+			// retrieve opcode from payload
+			opcode := payload[0] // opcode is first byte
+			// retrieve txid from payload
+			txidOffset := 4 // after 4 bytes, the txid starts
+			// the txid is 4 bytes, so we take four bytes after the offset
+			txid := payload[txidOffset : txidOffset+4]
+
+			c.logger.Info("Received packet", zap.Int("opCode", int(opcode)), zap.Any("transactionID", txid))
+			if opcode != dhcpOpCodeReply {
+				continue // opcode is not a reply, so continue
+			}
+
+			if TransactionID(txid) == xid {
+				break
+			}
+		}
+		// only occurs if we find our reply packet successfully
+		// a nil error means a reply was found for this txid
+		recvErrors <- nil
+	}(recvErrors)
+
+	select {
+	case err := <-recvErrors:
+		if err != nil {
+			return errors.Wrap(err, "error during receiving")
+		}
+	case <-time.After(DefaultTimeout):
```

Comment:
This leaks a goroutine if there's an error. Use `time.Ticker` and defer `time.(*Ticker).Stop()`,